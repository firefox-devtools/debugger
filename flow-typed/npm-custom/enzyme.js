/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// Using this custom definition of the enzyme module avoids generating some
// flow errors that are unrelated to uses of enzyme itself...
// It would be nice to clean this up.
declare module "enzyme" {
  declare module.exports: any;
}
