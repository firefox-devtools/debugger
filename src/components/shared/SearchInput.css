.search-field {
  width: calc(100% - 1px);
  height: 40px;
  background-color: var(--theme-body-background);
  border-bottom: 1px solid var(--theme-splitter-color);
  display: flex;
  flex-shrink: 0;
}

.search-field i {
  display: block;
  padding: 13px;
  width: 40px;
}

.search-field i svg {
  width: 16px;
}

.search-field input {
  border: none;
  line-height: 30px;
  font-size: 14px;
  background-color: var(--theme-body-background);
  color: var(--theme-comment);
  width: calc(100% - 38px);
  flex: 1;
}

.search-field .magnifying-glass {
  background-color: var(--theme-body-background);
}

.search-field .magnifying-glass path,
.search-field .magnifying-glass ellipse {
  stroke: var(--theme-splitter-color);
}

.search-field input::placeholder {
  color: var(--theme-body-color-inactive);
}

.search-field input:focus {
  outline-width: 0;
}

.search-field input.empty {
  color: var(--theme-highlight-orange);
}

.search-field .summary {
  line-height: 40px;
  padding-right: 10px;
  color: var(--theme-body-color-inactive);
}
