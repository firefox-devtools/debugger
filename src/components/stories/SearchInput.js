import React, { DOM as dom } from "react";
import { storiesOf, action } from "@kadira/storybook";
import _SearchInput from "../shared/SearchInput";
const SearchInput = React.createFactory(_SearchInput);
import { L10N } from "devtools-launchpad";
import { setValue } from "devtools-config";
import prefs from "../../utils/prefs";

import "../App.css";
import "../Editor/SearchBar.css";

import "devtools-launchpad/src/lib/themes/dark-theme.css";

// NOTE: we need this for supporting L10N in storybook
// we can move this to a shared helper as we add additional stories
if (typeof window == "object") {
  window.L10N = L10N;
  window.L10N.setBundle(require("../../../assets/panel/debugger.properties"));
}

function SearchInputFactory(options, { dir = "ltr", theme = "light" } = {}) {
  const themeClass = `theme-${theme}`;
  document.dir = dir;
  document.body.parentNode.className = themeClass;

  prefs.searchNav = true;

  return dom.div(
    {
      className: "",
      style: {
        width: "calc(100vw - 100px)",
        height: "calc(100vh - 100px)",
        margin: "auto",
        display: "flex",
        "flex-direction": "row"
      }
    },
    dom.div(
      {
        className: `search-bar ${themeClass}`,
        dir,
        style: {
          width: "100vw",
          "align-self": "center"
        }
      },
      SearchInput({
        ...{
          placeholder: "Search Input",
          query: "",
          handleNext: action("handleNext"),
          handlePrev: action("handlePrev")
        },
        ...options
      })
    )
  );
}

const stories = storiesOf("SearchInput", module);

const options = [{}, { dir: "rtl" }, { theme: "dark" }];
options.forEach(option => {
  const { dir, theme } = option;
  const optionLabel = dir || theme || "";
  stories
    .add(`simple ${optionLabel}`, () => {
      setValue("features.previewWatch", false);
      return SearchInputFactory({}, option);
    })
    .add(`no matches ${optionLabel}`, () => {
      setValue("features.previewWatch", false);
      return SearchInputFactory(
        {
          count: 0,
          query: "YO YO"
        },
        option
      );
    })
    .add(`matches ${optionLabel}`, () => {
      setValue("features.previewWatch", false);
      return SearchInputFactory(
        {
          count: 10,
          query: "yo"
        },
        option
      );
    });
});
