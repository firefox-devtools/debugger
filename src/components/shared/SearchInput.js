/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import React, { Component } from "react";

import { CloseButton } from "./Button";

import Svg from "./Svg";
import classnames from "classnames";
import "./SearchInput.css";

const arrowBtn = (onClick, type, className, tooltip) => {
  const props = {
    className,
    key: type,
    onClick,
    title: tooltip,
    type
  };

  return (
    <button {...props}>
      <Svg name={type} />
    </button>
  );
};

type Props = {
  count: number,
  expanded: boolean,
  handleClose?: (e: SyntheticMouseEvent<HTMLDivElement>) => void,
  handleNext?: (e: SyntheticMouseEvent<HTMLButtonElement>) => void,
  handlePrev?: (e: SyntheticMouseEvent<HTMLButtonElement>) => void,
  hasPrefix?: boolean,
  onBlur?: (e: SyntheticFocusEvent<HTMLInputElement>) => void,
  onChange: (e: SyntheticInputEvent<HTMLInputElement>) => void,
  onFocus?: (e: SyntheticFocusEvent<HTMLInputElement>) => void,
  onKeyDown: (e: SyntheticKeyboardEvent<HTMLInputElement>) => void,
  onKeyUp?: (e: SyntheticKeyboardEvent<HTMLInputElement>) => void,
  onHistoryScroll?: (historyValue: string) => void,
  placeholder: string,
  query: string,
  selectedItemId?: string,
  shouldFocus?: boolean,
  showErrorEmoji: boolean,
  size: string,
  summaryMsg: string,
  showClose: boolean
};

type State = {
  inputFocused: boolean,
  history: Array<string>
};

class SearchInput extends Component<Props, State> {
  displayName: "SearchInput";
  $input: ?HTMLInputElement;

  static defaultProps = {
    expanded: false,
    hasPrefix: false,
    selectedItemId: "",
    size: "",
    showClose: true
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      inputFocused: false,
      history: []
    };
  }

  componentDidMount() {
    this.setFocus();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.shouldFocus && !prevProps.shouldFocus) {
      this.setFocus();
    }
  }

  setFocus() {
    if (this.$input) {
      const input = this.$input;
      input.focus();

      if (!input.value) {
        return;
      }

      // omit prefix @:# from being selected
      const selectStartPos = this.props.hasPrefix ? 1 : 0;
      input.setSelectionRange(selectStartPos, input.value.length + 1);
    }
  }

  renderSvg() {
    const svgName = this.props.showErrorEmoji ? "sad-face" : "magnifying-glass";
    return <Svg name={svgName} />;
  }

  renderArrowButtons() {
    const { handleNext, handlePrev } = this.props;

    return [
      arrowBtn(
        handlePrev,
        "arrow-up",
        classnames("nav-btn", "prev"),
        L10N.getFormatStr("editor.searchResults.prevResult")
      ),
      arrowBtn(
        handleNext,
        "arrow-down",
        classnames("nav-btn", "next"),
        L10N.getFormatStr("editor.searchResults.nextResult")
      )
    ];
  }

  onFocus = (e: SyntheticFocusEvent<HTMLInputElement>) => {
    const { onFocus } = this.props;

    this.setState({ inputFocused: true });
    if (onFocus) {
      onFocus(e);
    }
  };

  onBlur = (e: SyntheticFocusEvent<HTMLInputElement>) => {
    const { onBlur } = this.props;

    this.setState({ inputFocused: false });
    if (onBlur) {
      onBlur(e);
    }
  };

  onKeyDown = (e: any) => {
    const { onHistoryScroll, onKeyDown } = this.props;
    if (!onHistoryScroll) {
      return onKeyDown(e);
    }

    const inputValue = e.target.value;
    const { history } = this.state;
    const currentHistoryIndex = history.indexOf(inputValue);

    if (e.key === "Enter") {
      this.saveEnteredTerm(inputValue);
      return onKeyDown(e);
    }

    if (e.key === "ArrowUp") {
      const previous =
        currentHistoryIndex > -1 ? currentHistoryIndex - 1 : history.length - 1;
      const previousInHistory = history[previous];
      if (previousInHistory) {
        e.preventDefault();
        onHistoryScroll(previousInHistory);
      }
      return;
    }

    if (e.key === "ArrowDown") {
      const next = currentHistoryIndex + 1;
      const nextInHistory = history[next];
      if (nextInHistory) {
        onHistoryScroll(nextInHistory);
      }
    }
  };

  saveEnteredTerm(query: string) {
    const { history } = this.state;
    const previousIndex = history.indexOf(query);
    if (previousIndex !== -1) {
      history.splice(previousIndex, 1);
    }
    history.push(query);
    this.setState({ history });
  }

  renderSummaryMsg() {
    const { summaryMsg } = this.props;

    if (!summaryMsg) {
      return null;
    }

    return <div className="summary">{summaryMsg}</div>;
  }

  renderNav() {
    const { count, handleNext, handlePrev } = this.props;
    if ((!handleNext && !handlePrev) || (!count || count == 1)) {
      return;
    }

    return (
      <div className="search-nav-buttons">{this.renderArrowButtons()}</div>
    );
  }

  render() {
    const {
      expanded,
      handleClose,
      onChange,
      onKeyUp,
      placeholder,
      query,
      selectedItemId,
      showErrorEmoji,
      size,
      showClose
    } = this.props;

    const inputProps = {
      className: classnames({
        empty: showErrorEmoji
      }),
      onChange,
      onKeyDown: e => this.onKeyDown(e),
      onKeyUp,
      onFocus: e => this.onFocus(e),
      onBlur: e => this.onBlur(e),
      "aria-autocomplete": "list",
      "aria-controls": "result-list",
      "aria-activedescendant":
        expanded && selectedItemId ? `${selectedItemId}-title` : "",
      placeholder,
      value: query,
      spellCheck: false,
      ref: c => (this.$input = c)
    };

    return (
      <div
        className={classnames("search-shadow", {
          focused: this.state.inputFocused
        })}
      >
        <div
          className={classnames("search-field", size)}
          role="combobox"
          aria-haspopup="listbox"
          aria-owns="result-list"
          aria-expanded={expanded}
        >
          {this.renderSvg()}
          <input {...inputProps} />
          {this.renderSummaryMsg()}
          {this.renderNav()}
          {showClose && (
            <CloseButton handleClick={handleClose} buttonClass={size} />
          )}
        </div>
      </div>
    );
  }
}

export default SearchInput;
