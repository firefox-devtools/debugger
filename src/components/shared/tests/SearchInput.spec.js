/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";

import SearchInput from "../SearchInput";

describe("SearchInput", () => {
  // !! wrapper is defined outside test scope
  // so it will keep values between tests
  const wrapper = shallow(
    <SearchInput
      query=""
      count={5}
      placeholder="A placeholder"
      summaryMsg="So many results"
      showErrorEmoji={false}
    />
  );

  it("renders", () => expect(wrapper).toMatchSnapshot());

  it("shows nav buttons", () => {
    wrapper.setProps({
      handleNext: jest.fn(),
      handlePrev: jest.fn()
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("shows svg error emoji", () => {
    wrapper.setProps({ showErrorEmoji: true });
    expect(wrapper).toMatchSnapshot();
  });

  it("shows svg magnifying glass", () => {
    wrapper.setProps({ showErrorEmoji: false });
    expect(wrapper).toMatchSnapshot();
  });
});
