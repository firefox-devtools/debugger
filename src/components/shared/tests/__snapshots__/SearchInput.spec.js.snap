// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput renders 1`] = `
<div
  className="search-shadow"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="result-list"
    className="search-field"
    role="combobox"
  >
    <Svg
      name="magnifying-glass"
    />
    <input
      aria-activedescendant=""
      aria-autocomplete="list"
      aria-controls="result-list"
      className=""
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="A placeholder"
      spellCheck={false}
      value=""
    />
    <div
      className="summary"
    >
      So many results
    </div>
    <CloseButton
      buttonClass=""
    />
  </div>
</div>
`;

exports[`SearchInput shows nav buttons 1`] = `
<div
  className="search-shadow"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="result-list"
    className="search-field"
    role="combobox"
  >
    <Svg
      name="magnifying-glass"
    />
    <input
      aria-activedescendant=""
      aria-autocomplete="list"
      aria-controls="result-list"
      className=""
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="A placeholder"
      spellCheck={false}
      value=""
    />
    <div
      className="summary"
    >
      So many results
    </div>
    <div
      className="search-nav-buttons"
    >
      <button
        className="nav-btn prev"
        key="arrow-up"
        onClick={[MockFunction]}
        title="Previous result"
        type="arrow-up"
      >
        <Svg
          name="arrow-up"
        />
      </button>
      <button
        className="nav-btn next"
        key="arrow-down"
        onClick={[MockFunction]}
        title="Next result"
        type="arrow-down"
      >
        <Svg
          name="arrow-down"
        />
      </button>
    </div>
    <CloseButton
      buttonClass=""
    />
  </div>
</div>
`;

exports[`SearchInput shows svg error emoji 1`] = `
<div
  className="search-shadow"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="result-list"
    className="search-field"
    role="combobox"
  >
    <Svg
      name="sad-face"
    />
    <input
      aria-activedescendant=""
      aria-autocomplete="list"
      aria-controls="result-list"
      className="empty"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="A placeholder"
      spellCheck={false}
      value=""
    />
    <div
      className="summary"
    >
      So many results
    </div>
    <div
      className="search-nav-buttons"
    >
      <button
        className="nav-btn prev"
        key="arrow-up"
        onClick={[MockFunction]}
        title="Previous result"
        type="arrow-up"
      >
        <Svg
          name="arrow-up"
        />
      </button>
      <button
        className="nav-btn next"
        key="arrow-down"
        onClick={[MockFunction]}
        title="Next result"
        type="arrow-down"
      >
        <Svg
          name="arrow-down"
        />
      </button>
    </div>
    <CloseButton
      buttonClass=""
    />
  </div>
</div>
`;

exports[`SearchInput shows svg magnifying glass 1`] = `
<div
  className="search-shadow"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="result-list"
    className="search-field"
    role="combobox"
  >
    <Svg
      name="magnifying-glass"
    />
    <input
      aria-activedescendant=""
      aria-autocomplete="list"
      aria-controls="result-list"
      className=""
      onBlur={[Function]}
      onFocus={[Function]}
      placeholder="A placeholder"
      spellCheck={false}
      value=""
    />
    <div
      className="summary"
    >
      So many results
    </div>
    <div
      className="search-nav-buttons"
    >
      <button
        className="nav-btn prev"
        key="arrow-up"
        onClick={[MockFunction]}
        title="Previous result"
        type="arrow-up"
      >
        <Svg
          name="arrow-up"
        />
      </button>
      <button
        className="nav-btn next"
        key="arrow-down"
        onClick={[MockFunction]}
        title="Next result"
        type="arrow-down"
      >
        <Svg
          name="arrow-down"
        />
      </button>
    </div>
    <CloseButton
      buttonClass=""
    />
  </div>
</div>
`;
